python/stempy/image/__init__.py

```diff
@@ -653,3 +653,116 @@ def _electron_counted_metadata_to_dict(metadata):
+    :rype plot: bool
+    
+    :rtype: np.ndarray
+    
+    :example:
+    >>> sp = stempy.io.load_electron_counts('file.h5')
+    >>> dp2 = stempy.image.virtual_darkfield(sp, (288, 260), (288, 160), (10, 10)) # 2 apertures
+    >>> dp1 = stempy.image.virtual_darkfield(sp, 260, 160, 10) # 1 aperture
+    
+    """
+    
+    # Change to iterable if single value
+    if isinstance(centers_x, (int, float)):
+         centers_x = (centers_x,)
+    if isinstance(centers_y, (int, float)):
+         centers_y = (centers_y,)
+    if isinstance(radii, (int, float)):
+         radii = (radii,)
+
+    rs_image = np.zeros((array.shape[0] * array.shape[1],), dtype=array.dtype)
+    for ii, events in enumerate(array.data):
+        for ev in events:
+            ev_rows = ev // array.frame_shape[0]
+            ev_cols = ev % array.frame_shape[1]
+            for cc_0, cc_1, rr in zip(centers_x, centers_y, radii):
+                dist = np.sqrt((ev_rows - cc_1)**2 + (ev_cols - cc_0)**2)
+                rs_image[ii] += len(np.where(dist < rr)[0])
+    rs_image = rs_image.reshape(array.shape[0:2])
+    
+    return rs_image
+
+def plot_virtual_darkfield(image, centers_x, centers_y, radii, axes=None):
+    """Plot circles on the diffraction pattern corresponding to the position and size of virtual dark field apertures.
+    This has the same input as stempy.image.virtual_darkfield so users can check their input is physically correct.
+    
+    :param image: The diffraciton pattern to plot over
+    :type image: np.ndarray, 2D
+    
+    :param centers_x: The center of each round aperture as the row locations
+    :type centers_x: iterable
+    
+    :param centers_y: The center of each round aperture as the column locations
+    :type centers_y: iterable
+    
+    :param radii: The radius of each aperture. This has to be in the form (r0, )
+    :type radii: iterable
+    
+    :param axes: A matplotlib axes instance to use for the plotting. If None then a new plot is created.
+    :type axes: matplotlib.axes._subplots.AxesSubplot
+    
+    :rtype: matplotlib.axes._subplots.AxesSubplot
+    """
+    import matplotlib.pyplot as plt
+    from matplotlib.colors import LogNorm
+    from matplotlib.patches import Circle
+
+    # Change to iterable if single value
+    if isinstance(centers_x, (int, float)):
+         centers_x = (centers_x,)
+    if isinstance(centers_y, (int, float)):
+         centers_y = (centers_y,)
+    if isinstance(radii, (int, float)):
+         radii = (radii,)
+    
+    if not axes:
+        fg, axes = plt.subplots(1, 1)
+    
+    axes.imshow(image, cmap='magma', norm=LogNorm())
+    
+    # Place a circle at each apertue location
+    for cc_0, cc_1, rr in zip(centers_x, centers_y, radii):
+        C = Circle((cc_0, cc_1), rr, fc='none', ec='c')
+        axes.add_patch(C)
+    
+    return axes
+
+def mask_real_space(array, mask):
```

Comment:
Nice! This can potentially also be a method on the `SparseArray` itself.

Reply:
I agree. This sort of allows for "fancy indexing". Maybe the same algorithm can be used for that functionality.